File: backend/src/modules/job/jobService.js
import { prisma } from '../../config/db.js';

function withAvailableSlot(job) {
  return { ...job, availableSlot: Math.max(job.totalSlot - job.filledSlot, 0) };
}

export async function createJob(employerId, data) {
  const {
    title, description, locationArea, fullAddress,
    totalSlot, salaryAmount, salaryType, requirements,
  } = data;

  if (!title || !description || !locationArea || !fullAddress || !salaryAmount) {
    throw { status: 400, message: 'Semua field wajib diisi, termasuk lokasi dan gaji' };
  }

  return prisma.jobPosting.create({
    data: {
      employerId,
      title,
      description,
      locationArea,
      fullAddress,
      totalSlot: totalSlot ? Number(totalSlot) : 1,
      salaryAmount: Number(salaryAmount),
      salaryType: salaryType || 'PER_HARI',
      requirements: requirements || [],
    },
  });
}

export async function updateJob(employerId, jobId, data) {
  const job = await prisma.jobPosting.findUnique({ where: { id: jobId } });

  if (!job) throw { status: 404, message: 'Lowongan tidak ditemukan' };
  if (job.employerId !== employerId) throw { status: 403, message: 'Anda tidak berhak mengubah lowongan ini' };

  const {
    title, description, locationArea, fullAddress, status,
    salaryAmount, salaryType, requirements,
  } = data;

  return prisma.jobPosting.update({
    where: { id: jobId },
    data: {
      ...(title && { title }),
      ...(description && { description }),
      ...(locationArea && { locationArea }),
      ...(fullAddress && { fullAddress }),
      ...(status && { status }),
      ...(salaryAmount && { salaryAmount: Number(salaryAmount) }),
      ...(salaryType && { salaryType }),
      ...(requirements && { requirements }),
    },
  });
}

export async function deleteJob(employerId, jobId) {
  const job = await prisma.jobPosting.findUnique({ where: { id: jobId } });

  if (!job) throw { status: 404, message: 'Lowongan tidak ditemukan' };
  if (job.employerId !== employerId) throw { status: 403, message: 'Anda tidak berhak menghapus lowongan ini' };

  await prisma.jobPosting.delete({ where: { id: jobId } });
  return { message: 'Lowongan berhasil dihapus' };
}

export async function searchJobs({ page = 1, limit = 10, locationArea, keyword }) {
  const skip = (page - 1) * limit;

  const where = {
    status: 'OPEN',
    ...(locationArea && { locationArea: { contains: locationArea } }),
    ...(keyword && { title: { contains: keyword } }),
  };

  const [items, total] = await Promise.all([
    prisma.jobPosting.findMany({
      where,
      skip,
      take: Number(limit),
      orderBy: { createdAt: 'desc' },
      include: {
        employer: { select: { id: true, profile: { select: { fullName: true, photoUrl: true } } } },
      },
    }),
    prisma.jobPosting.count({ where }),
  ]);

  return {
    items: items.map(withAvailableSlot),
    pagination: { page: Number(page), limit: Number(limit), total, totalPages: Math.ceil(total / limit) },
  };
}

export async function getMyJobs(employerId) {
  const jobs = await prisma.jobPosting.findMany({
    where: { employerId },
    orderBy: { createdAt: 'desc' },
    include: { _count: { select: { applications: true } } },
  });

  return jobs.map(withAvailableSlot);
}

export async function getRecommendedJobs(workerId, { page = 1, limit = 10 }) {
  const profile = await prisma.profile.findUnique({ where: { userId: workerId } });
  const skills = profile?.skills;

  if (!Array.isArray(skills) || skills.length === 0) {
    throw { status: 400, message: 'Anda belum mengisi keahlian di profil, silakan lengkapi dulu' };
  }

  const textMatchConditions = skills.flatMap((skill) => ([
    { title: { contains: skill } },
    { description: { contains: skill } },
  ]));

  const [items, total] = await Promise.all([
    prisma.jobPosting.findMany({
      where: { status: 'OPEN', OR: textMatchConditions },
      skip: (page - 1) * limit,
      take: Number(limit),
      orderBy: { createdAt: 'desc' },
      include: { employer: { select: { id: true, profile: { select: { fullName: true, photoUrl: true } } } } },
    }),
    prisma.jobPosting.count({ where: { status: 'OPEN', OR: textMatchConditions } }),
  ]);

  return {
    items: items.map(withAvailableSlot),
    pagination: { page: Number(page), limit: Number(limit), total, totalPages: Math.ceil(total / limit) },
  };
}
export async function getJobDetail(jobId, currentUserId) {
  const job = await prisma.jobPosting.findUnique({
    where: { id: jobId },
    include: {
      employer: {
        select: { id: true, profile: { select: { fullName: true, photoUrl: true, location: true } } },
      },
    },
  });

  if (!job) throw { status: 404, message: 'Lowongan tidak ditemukan' };

  const employerRatingAgg = await prisma.rating.aggregate({
    where: { toUserId: job.employerId },
    _avg: { score: true },
    _count: { score: true },
  });

  let myApplication = null;
  if (currentUserId) {
    myApplication = await prisma.application.findUnique({
      where: { jobId_applicantId: { jobId, applicantId: currentUserId } },
      select: { id: true, status: true, type: true },
    });
  }

  return {
    ...withAvailableSlot(job),
    employer: {
      ...job.employer,
      rating: {
        average: employerRatingAgg._avg.score ? Math.round(employerRatingAgg._avg.score * 10) / 10 : 0,
        total: employerRatingAgg._count.score,
      },
    },
    myApplication,
  };
}

File: frontend/lib/services/job_service.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'api_client.dart';
import 'auth_service.dart' show ApiException;

class JobService {
  static Future<Map<String, dynamic>> searchJobs({
    String? keyword,
    String? locationArea,
    int page = 1,
    int limit = 10,
  }) async {
    final queryParams = {
      'page': '$page',
      'limit': '$limit',
      if (keyword != null && keyword.isNotEmpty) 'keyword': keyword,
      if (locationArea != null && locationArea.isNotEmpty)
        'locationArea': locationArea,
    };
    final uri = Uri.parse('${ApiConfig.baseUrl}/jobs')
        .replace(queryParameters: queryParams);
    final headers = await ApiConfig.authHeaders();

    final response = await http.get(uri, headers: headers);
    final data = jsonDecode(response.body) as Map<String, dynamic>;

    if (response.statusCode != 200) {
      throw ApiException(data['message'] ?? 'Gagal memuat lowongan',
          statusCode: response.statusCode);
    }
    return data; // { items, pagination }
  }

  static Future<Map<String, dynamic>> getJobDetail(String jobId) async {
    final uri = Uri.parse('${ApiConfig.baseUrl}/jobs/$jobId');
    final headers = await ApiConfig.authHeaders();

    final response = await http.get(uri, headers: headers);
    final data = jsonDecode(response.body) as Map<String, dynamic>;

    if (response.statusCode != 200) {
      throw ApiException(data['message'] ?? 'Gagal memuat detail lowongan',
          statusCode: response.statusCode);
    }
    return data;
  }

  static Future<List<dynamic>> getMyJobs() async {
    final uri = Uri.parse('${ApiConfig.baseUrl}/jobs/mine');
    final headers = await ApiConfig.authHeaders();

    final response = await http.get(uri, headers: headers);
    final decoded = jsonDecode(response.body);

    if (response.statusCode != 200) {
      final message = decoded is Map ? decoded['message'] : null;
      throw ApiException(message ?? 'Gagal memuat lowongan saya',
          statusCode: response.statusCode);
    }
    return decoded as List<dynamic>;
  }

  static Future<Map<String, dynamic>> createJob({
    required String title,
    required String description,
    required String locationArea,
    required String fullAddress,
    required num salaryAmount,
    String salaryType = 'PER_HARI',
    int totalSlot = 1,
    List<String> requirements = const [],
  }) async {
    final uri = Uri.parse('${ApiConfig.baseUrl}/jobs');
    final headers = await ApiConfig.authHeaders();

    final response = await http.post(
      uri,
      headers: headers,
      body: jsonEncode({
        'title': title,
        'description': description,
        'locationArea': locationArea,
        'fullAddress': fullAddress,
        'salaryAmount': salaryAmount,
        'salaryType': salaryType,
        'totalSlot': totalSlot,
        'requirements': requirements,
      }),
    );

    final data = jsonDecode(response.body) as Map<String, dynamic>;

    if (response.statusCode != 201) {
      throw ApiException(data['message'] ?? 'Gagal membuat lowongan',
          statusCode: response.statusCode);
    }
    return data;
  }

  static Future<Map<String, dynamic>> updateJob({
    required String jobId,
    String? title,
    String? description,
    String? locationArea,
    String? fullAddress,
    num? salaryAmount,
    String? salaryType,
    List<String>? requirements,
    String? status,
  }) async {
    final uri = Uri.parse('${ApiConfig.baseUrl}/jobs/$jobId');
    final headers = await ApiConfig.authHeaders();

    final body = <String, dynamic>{
      if (title != null) 'title': title,
      if (description != null) 'description': description,
      if (locationArea != null) 'locationArea': locationArea,
      if (fullAddress != null) 'fullAddress': fullAddress,
      if (salaryAmount != null) 'salaryAmount': salaryAmount,
      if (salaryType != null) 'salaryType': salaryType,
      if (requirements != null) 'requirements': requirements,
      if (status != null) 'status': status,
    };

    final response =
        await http.put(uri, headers: headers, body: jsonEncode(body));
    final data = jsonDecode(response.body) as Map<String, dynamic>;

    if (response.statusCode != 200) {
      throw ApiException(data['message'] ?? 'Gagal menyimpan perubahan',
          statusCode: response.statusCode);
    }
    return data;
  }

  static Future<void> deleteJob(String jobId) async {
    final uri = Uri.parse('${ApiConfig.baseUrl}/jobs/$jobId');
    final headers = await ApiConfig.authHeaders();

    final response = await http.delete(uri, headers: headers);
    final data = jsonDecode(response.body) as Map<String, dynamic>;

    if (response.statusCode != 200) {
      throw ApiException(data['message'] ?? 'Gagal menghapus lowongan',
          statusCode: response.statusCode);
    }
  }
}
